shanoir-ng-front/src/app/shared/mass-download/mass-download.service.ts

```diff
@@ -0,0 +1,492 @@
+                return this.openAltModal().then(ret => {
+                    if (ret != 'cancel') {
+                        return this._downloadByIdsAlt(datasetIds, ret);
+                    } else return Promise.resolve();
+                });
+            } else throw error;
+        });
+    }
+
+    downloadDatasets(datasets: Dataset[]): Promise<void> {
+        return this.openModal().then(ret => {
+            if (ret != 'cancel') {
+                return this._downloadDatasets(datasets, ret.format, ret.nbQueues);
+            } else return Promise.resolve();
+        });
+    }
+
+    downloadAllByStudyId(studyId: number): Promise<void> {
+        return this.openModal().then(ret => {
+            if (ret != 'cancel') {
+                return this.datasetService.getByStudyId(studyId).then(datasets => {
+                    this._downloadDatasets(datasets, ret.format, ret.nbQueues);
+                })
+            } else return Promise.resolve();
+        });
+    }
+
+    downloadAllByStudyIdAndSubjectId(studyId: number, subjectId: number): Promise<void> {
+        return this.openModal().then(ret => {
+            if (ret != 'cancel') {
+                return this.datasetService.getByStudyIdAndSubjectId(studyId, subjectId).then(datasets => {
+                    return this._downloadDatasets(datasets, ret.format, ret.nbQueues).then(r => {
+                        return r;
+                    });
+                })
+            } else return Promise.resolve();
+        });
+    }
+
+    private _downloadByIds(datasetIds: number[], format: Format, nbQueues: number = 4, task?: Task, report?: Report, parentHandle?: FileSystemDirectoryHandle): Promise<void> {
+        if (datasetIds.length == 0) return;
+        let directoryHandlePromise: Promise<FileSystemDirectoryHandle>;
+        if (parentHandle) {
+            directoryHandlePromise = Promise.resolve(parentHandle);
+        } else {
+            directoryHandlePromise = this.getFolderHandle();
+        }
+        return directoryHandlePromise.then(parentFolderHandle => { // ask the user's parent directory
+            if (!task) task = this.createTask(datasetIds.length);
+            return this.downloadQueue.waitForTurn().then(releaseQueue => {
+                try {
+                    task.status = 2;
+                    task.lastUpdate = new Date();
+                    this.notificationService.pushLocalTask(task);
+                    const start: number = Date.now();
+                    let ids = [...datasetIds]; // copy array
+                    if (!report) report = this.initReport(datasetIds, task.id, parentFolderHandle.name, format, nbQueues);
+                    let promises: Promise<void>[] = [];
+                    for (let queueIndex = 0; queueIndex < nbQueues; queueIndex++) { // build the dl queues
+                        promises.push(
+                            this.recursiveSave(ids.shift(), format, parentFolderHandle, ids, report, task)
```

Comment:
I'm not really sure but here, if you have 4 queues and 8 datasets
ids [1,2,3,4,5,6,7,8]
Q1:(1, ... [1,2,3,4,5,6,7,8]);
Q2:(2, ... [2,3,4,5,6,7,8]);
Q3:(3, ... [3,4,5,6,7,8]);
Q4:(4, ... [4,5,6,7,8]);
Promises are executed as soon as declared, so only Q1 would work no? I think I need clarifications about this ;)